tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/mailet/CalDavCollect.java

```diff
@@ -104,14 +107,46 @@ private Mono<Void> synchronizeWithDavServer(Calendar calendar, DavUser davUser)
             return deleteDavCalendar(davUser, eventUid);
         } else {
             return davClient.getCalendarObject(davUser, eventUid)
-                .flatMap(davCalendarObject -> updateDavCalendar(davCalendarObject, calendar, davUser))
-                .switchIfEmpty(createDavCalendar(calendar, davUser));
+                .singleOptional()
+                .flatMap(maybeDavCalendarObject ->
+                    maybeDavCalendarObject.map(davCalendarObject -> updateDavCalendar(davCalendarObject, calendar, davUser))
+                        .orElseGet(() -> createDavCalendar(calendar, davUser)));
         }
     }
 
-    private Mono<Void> updateDavCalendar(DavCalendarObject davCalendarObject, Calendar calendar, DavUser davUser) {
-        // TODO Rewrite this method to handle case the calendar event is already present
-        return Mono.error(new RuntimeException("Calendar event already exists"));
+    private Mono<Void> updateDavCalendar(DavCalendarObject davCalendarObject, Calendar newCalendar, DavUser davUser) {
+        if (checkIfNewCalendarIsNewVersion(newCalendar, davCalendarObject.calendarData())) {
+            DavCalendarObject newDavCalendarObject = new DavCalendarObject(davCalendarObject.uri(),
+                newCalendar.removeIf(property -> Property.METHOD.equals(property.getName())),
+                davCalendarObject.eTag());
+            return davClient.doUpdateCalendarObject(davUser.username(), newDavCalendarObject);
+        } else {
+            return Mono.empty();
+        }
+    }
+
+    private boolean checkIfNewCalendarIsNewVersion(Calendar newCalendar, Calendar currentCalendar) {
+        CalendarComponent newVEvent = newCalendar.getComponent(Component.VEVENT).get();
+        CalendarComponent currentVEvent = currentCalendar.getComponent(Component.VEVENT).get();
+        Optional<Property> maybeNewSequence = newVEvent.getProperty(Property.SEQUENCE);
+        Optional<Property> maybeCurrentSequence = currentVEvent.getProperty(Property.SEQUENCE);
+        if (maybeNewSequence.isPresent() && maybeCurrentSequence.isPresent()) {
+            int newSequence = Integer.parseInt(maybeNewSequence.get().getValue());
+            int currentSequence = Integer.parseInt(maybeCurrentSequence.get().getValue());
+            if (newSequence > currentSequence) {
+                return true;
+            } else if (newSequence == currentSequence) {
+                return checkIfNewDtStampIsOlder(newVEvent, currentVEvent);
+            } else {
+                return false;
+            }
+        } else {
+            return checkIfNewDtStampIsOlder(newVEvent, currentVEvent);
+        }
```

Comment:
IMO those boolean conditions are complex and hard to read.

It defeats functional constructs.

Let me propose:

```
        Optional<Integer> maybeNewSequence = newVEvent.getProperty(Property.SEQUENCE).map(Integer::parseInt);
        Optional<Property> maybeCurrentSequence = currentVEvent.getProperty(Property.SEQUENCE).map(Integer::parseInt);
        maybeNewSequence.flapMap(newSequence -> maybeCurrentSequence
            .filter(currentSequence -> currentSequence != newSequence)
            .map(currentSequence -> currentSequence > newSequence))
    .getOrElse(() -> hasNewerDtstamp(newCalendar, currentCalendar);
```